plugins/text-editor-resources/src/components/Popup.svelte

```diff
@@ -0,0 +1,122 @@
+<!--
+// Copyright © 2020, 2021 Anticrm Platform Contributors.
+// Copyright © 2022, 2023, 2024 Hardcore Engineering Inc.
+//
+// Licensed under the Eclipse Public License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License. You may
+// obtain a copy of the License at https://www.eclipse.org/legal/epl-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+//
+// See the License for the specific language governing permissions and
+// limitations under the License.
+-->
+<script lang="ts">
+  import { deviceOptionsStore as deviceInfo, PopupResult, resizeObserver, showPopup } from '@hcengineering/ui'
+  import { onDestroy, onMount } from 'svelte'
+  import DummyPopup from './DummyPopup.svelte'
+
+  export let query: string = ''
+  export let clientRect: () => ClientRect
+  export let command: (props: any) => void
+  export let close: () => void
+
+  let popup: HTMLDivElement
+  let dummyPopup: PopupResult
+
+  onMount(() => {
+    dummyPopup = showPopup(
+      DummyPopup,
+      {},
+      undefined,
+      () => {
+        close()
+        command(null)
+      },
+      () => {},
+      { overlay: false, category: '' }
+    )
+  })
+
+  onDestroy(() => {
+    dummyPopup.close()
+  })
+
+  export function done(): void {}
+
+  function updateStyle(): void {
+    const rect = clientRect()
+    const wDoc = $deviceInfo.docWidth
+    const hDoc = $deviceInfo.docHeight
+    let tempStyle = ''
+    if (rect.top < hDoc - rect.bottom) {
```

Comment:
Just out of curiosity, cannot we turn this into [decompose conditional](https://refactoring.guru/decompose-conditional)?

Same for the other `if else` blocks.

Reply:
Yes I think so, this was an older component `MentionList.svelte`. I slitted in two to share the logic between this and `EmojiList.svelte` component